de.hamstersimulator.objectsfirst.inspector/src/main/java/de/hamstersimulator/objectsfirst/inspector/ui/InputControl.java

```diff
@@ -34,124 +47,234 @@ public class InputControl extends HBox {
+     * Property if the current input is valid
+     *
+     * @return a readonly property for isValid
+     */
+    public ReadOnlyBooleanProperty isValidProperty() {
+        return this.isValid.getReadOnlyProperty();
     }
 
+    /**
+     * Property if the user can update the value of this control
+     *
+     * @return a property for isReadOnly
+     */
     public BooleanProperty isReadOnlyProperty() {
         return this.isReadOnly;
     }
 
-    private void handleValueChanged(Object value) {
-        if (!this.changeFromInputControl) {
-            if (this.type.getCategory() == TypeCategory.OPTIONAL) {
-                assert value instanceof Optional;
-                value = ((Optional<?>)value).orElse(null);
+    /**
+     * Inits the change listeners which watch for changes on the value and
+     * the instance list of the InspectionViewModel.
+     * Should only be called once from the constructor
+     */
+    private void initChangeListeners() {
+        this.value.addListener((observable, oldValue, newValue) -> {
+            this.handleValueChanged(newValue);
+        });
+        this.inspectionViewModel.instancesProperty().addListener((ListChangeListener<InstanceViewModel>) change -> {
+            if (this.instanceViewModelListChangeListener != null) {
+                this.instanceViewModelListChangeListener.onChanged(change);
             }
-            if (value != null
-                    && (this.type.getCategory() == TypeCategory.OBJECT || this.type.getCategory() == TypeCategory.OPTIONAL)) {
-                final Type newType = new Type(value.getClass());
-                final TypeCategory newCategory = newType.getCategory();
-                final TypeCategory currentCategory = this.currentType.get().getCategory();
-                if (newCategory == TypeCategory.ENUM && !this.currentType.get().getType().isAssignableFrom(newType.getType())) {
-                    boolean foundMatchingEnumType = false;
-                    for (final Type type : this.inspectionViewModel.multiInputTypesProperty()) {
-                        if (type.getCategory() == TypeCategory.ENUM && type.getType().isAssignableFrom(newType.getType())) {
-                            this.multiTypeComboBox.setValue(type);
-                            foundMatchingEnumType = true;
-                            break;
-                        }
-                    }
-                    if (!foundMatchingEnumType) {
-                        this.multiTypeComboBox.setValue(Type.OBJECT_TYPE);
-                    }
-                }
-                if (newCategory != currentCategory && currentCategory != TypeCategory.OBJECT) {
-                    this.multiTypeComboBox.setValue(newType);
-                }
+        });
+    }
+
+    /**
+     * Creates the children for the complex UI which is displayed if all input
+     * types must be possible (OBJECT and OPTIONAL)
+     *
+     * @return The ComboBox used too switch between the different
```

Comment:
to instead of too

Reply:
fixed with d5ae83d9b6d00633c06a658d67118784234e9597